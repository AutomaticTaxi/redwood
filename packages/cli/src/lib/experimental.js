import path from 'path'

import { getConfig, getPaths, resolveFile } from '@redwoodjs/project-config'

/**
 * Checks to see if the user has the experimental server file present at api/src/server.ts|js and enabled in redwood.toml
 *
 * @returns {boolean} true if the experimental server file is enabled and present
 */
export function isUsingExperimentalServerFile() {
  return (
    getConfig().experimental?.serverFile?.enabled &&
    resolveFile(path.join(getPaths().api.src, 'server'))
  )
}
