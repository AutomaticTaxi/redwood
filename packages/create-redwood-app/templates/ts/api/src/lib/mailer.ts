import { Mailer } from '@redwoodjs/mailer-core'
import { InMemoryMailHandler } from '@redwoodjs/mailer-handler-in-memory'
import { NodemailerMailHandler } from '@redwoodjs/mailer-handler-nodemailer'
import { StudioMailHandler } from '@redwoodjs/mailer-handler-studio'
import { ReactEmailRenderer } from '@redwoodjs/mailer-renderer-react-email'

import { logger } from './logger'

export const mailer = new Mailer({
  // Handlers to send our emails out to the world
  handling: {
    handlers: {
      inMemory: new InMemoryMailHandler(),
      studio: new StudioMailHandler(),
      // TODO: Update this handler config or switch it out for a different handler completely
      nodemailer: new NodemailerMailHandler({
        transport: {
          host: 'localhost',
          port: 4319,
          secure: false,
        },
      }),
    },
    // This is our default handler when in production only
    default: 'nodemailer',
  },

  // Renderers to convert our templates into HTML/text
  rendering: {
    renderers: {
      reactEmail: new ReactEmailRenderer(),
    },
    default: 'reactEmail',
  },

  // Specific handlers for testing and development time
  test: {
    handler: 'inMemory',
    // when: process.env.NODE_ENV === 'test',
  },
  development: {
    handler: 'studio',
    // when: process.env.NODE_ENV !== 'production',
  },

  // Default send options
  defaults: {
    // replyTo: 'noreply@example.com',
  },

  logger,
})
