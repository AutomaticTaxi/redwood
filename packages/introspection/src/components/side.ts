import fs from 'node:fs'

import { getBaseDir, getConfigPath, getPaths } from '@redwoodjs/project-config'

import type {
  RedwoodIntrospectionError,
  RedwoodIntrospectionWarning,
} from './introspection'
import { RedwoodIntrospectionComponent } from './introspection'

export type RedwoodSideType = 'web' | 'api' | 'unknown'

export class RedwoodSide extends RedwoodIntrospectionComponent {
  readonly type = 'side'

  readonly sideType: RedwoodSideType

  private constructor(filepath: string) {
    super(filepath)

    switch (this.name) {
      case 'web':
        this.sideType = 'web'
        break
      case 'api':
        this.sideType = 'api'
        break
      default:
        this.sideType = 'unknown'
        break
    }
  }

  getErrors(): RedwoodIntrospectionError[] {
    const errors: RedwoodIntrospectionError[] = []

    // Check if the side is supported - currently only 'web' and 'api' are supported
    if (this.sideType === 'unknown') {
      errors.push({
        component: {
          type: this.type,
          filepath: this.filepath,
          name: this.name,
        },
        message: `Side from '${this.name}' is not supported`,
      })
    }

    return errors
  }

  getWarnings(): RedwoodIntrospectionWarning[] {
    // No warnings for sides
    return []
  }

  static parseSide(filepath: string) {
    return new RedwoodSide(filepath)
  }

  static parseSides(directory: string = getPaths().base) {
    const sides: RedwoodSide[] = []

    // In the future we should update this to handle whatever dynamic/multiple side support we have
    const paths = getPaths(getBaseDir(getConfigPath(directory)))
    if (fs.existsSync(paths.web.base)) {
      sides.push(new RedwoodSide(paths.web.base))
    }
    if (fs.existsSync(paths.api.base)) {
      sides.push(new RedwoodSide(paths.api.base))
    }

    return sides
  }
}
