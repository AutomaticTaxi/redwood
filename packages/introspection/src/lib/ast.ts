import fs from 'node:fs'
import path from 'node:path'

import { parse as babelParse } from '@babel/parser'
import type { ParserPlugin } from '@babel/parser'
import traverse from '@babel/traverse'
import type {
  JSXElement,
  ExportNamedDeclaration,
  File,
  VariableDeclarator,
  ExportDefaultDeclaration,
} from '@babel/types'
import {
  assertJSXOpeningElement,
  assertJSXIdentifier,
  assertIdentifier,
  isJSXAttribute,
  isJSXSpreadAttribute,
  isVariableDeclaration,
  isVariableDeclarator,
  isIdentifier,
} from '@babel/types'

import { getPaths } from '@redwoodjs/project-config'

export function getASTFromFile(filepath: string) {
  const code = fs.readFileSync(filepath, { encoding: 'utf8', flag: 'r' })

  // use jsx plugin for web files, because in JS, the .jsx extension is not used
  const isJsxFile =
    path.extname(filepath).match(/[jt]sx$/) ||
    path.parse(filepath).dir.startsWith(getPaths().web.base)

  const plugins = [
    'typescript',
    'nullishCoalescingOperator',
    'objectRestSpread',
    isJsxFile && 'jsx',
  ].filter(Boolean) as ParserPlugin[]

  return babelParse(code, {
    sourceType: 'module',
    plugins,
  })
}

export function getSpecificNamedExportDeclarations(ast: File, names: string[]) {
  const desiredNamedExports: Map<string, VariableDeclarator> = new Map()
  const allNamedExports: ExportNamedDeclaration[] = []
  traverse(ast, {
    ExportNamedDeclaration: (path) => {
      allNamedExports.push(path.node)
    },
  })
  allNamedExports.forEach((namedExport) => {
    if (
      namedExport.declaration != null &&
      isVariableDeclaration(namedExport.declaration) &&
      isVariableDeclarator(namedExport.declaration.declarations[0]) &&
      isIdentifier(namedExport.declaration.declarations[0].id) &&
      names.includes(namedExport.declaration.declarations[0].id.name)
    ) {
      desiredNamedExports.set(
        namedExport.declaration.declarations[0].id.name,
        namedExport.declaration.declarations[0]
      )
    }
  })
  return desiredNamedExports
}

export function getExportDefaultDeclaration(ast: File) {
  let defaultExport: ExportDefaultDeclaration | undefined
  traverse(ast, {
    ExportDefaultDeclaration: (path) => {
      defaultExport = path.node
    },
  })
  return defaultExport
}

export function getVariableDeclarator(ast: File, name: string) {
  let variableDeclarator: VariableDeclarator | undefined
  traverse(ast, {
    VariableDeclarator: (path) => {
      if (isIdentifier(path.node.id) && path.node.id.name === name) {
        variableDeclarator = path.node
      }
    },
  })
  return variableDeclarator
}

export function getJSXElementAttributes(element: JSXElement) {
  const attributes: Map<string, string> = new Map()
  assertJSXOpeningElement(element.openingElement)
  if (element.openingElement.attributes.length > 0) {
    element.openingElement.attributes.forEach((attribute) => {
      if (isJSXAttribute(attribute)) {
        assertJSXIdentifier(attribute.name)
        const key = attribute.name.name
        let value: string
        switch (attribute.value?.type) {
          case 'StringLiteral':
            value = attribute.value.value
            break
          case 'JSXExpressionContainer':
            assertIdentifier(attribute.value.expression)
            value = attribute.value.expression.name
            break
          case undefined: // Accept this because it could be a attribute with no value, i.e. notfound
            value = 'true'
            break
          default:
            throw new Error(
              `Unknown value type found while trying to process the ${key} property when extracting all attributes`
            )
        }
        attributes.set(key, value)
      } else if (isJSXSpreadAttribute(attribute)) {
        throw new Error(
          'JSXSpreadAttribute type is not yet supported when extracting all attributes'
        )
      } else {
        throw new Error(
          'Uknown attribute type found while trying to extract all attributes'
        )
      }
    })
  }
  return attributes
}
