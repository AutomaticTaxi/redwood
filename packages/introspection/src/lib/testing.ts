import { ensurePosixPath } from '@redwoodjs/project-config'

export function stripAndFormatPathForTesting(
  filepath: string,
  projectFilepath: string
) {
  if (filepath.startsWith(projectFilepath)) {
    filepath = ensurePosixPath(filepath.substring(projectFilepath.length))
  }
  return filepath
}
