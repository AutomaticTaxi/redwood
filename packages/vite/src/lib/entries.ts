import path from 'node:path'

import type { PagesDependency } from '@redwoodjs/project-config'
import {
  ensurePosixPath,
  getPaths,
  processPagesDir,
} from '@redwoodjs/project-config'

const getPathRelativeToSrc = (maybeAbsolutePath: string) => {
  // If the path is already relative
  if (!path.isAbsolute(maybeAbsolutePath)) {
    return maybeAbsolutePath
  }

  return `./${path.relative(getPaths().web.src, maybeAbsolutePath)}`
}

const withRelativeImports = (page: PagesDependency) => {
  return {
    ...page,
    relativeImport: ensurePosixPath(getPathRelativeToSrc(page.importPath)),
  }
}

export function getEntries() {
  return processPagesDir()
    .map(withRelativeImports)
    .reduce(
      (acc, page) => {
        acc[page.importName] = page.path
        return acc
      },
      {} as Record<string, string>,
    )
}

export async function getEntriesFromDist(): Promise<Record<string, string>> {
  const entriesDist = getPaths().web.distRscEntries
  const { serverEntries } = await import(`file://${entriesDist}`)
  return serverEntries
}
