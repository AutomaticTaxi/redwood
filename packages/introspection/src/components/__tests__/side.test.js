import path from 'node:path'

import { stripAndFormatPathForTesting } from '../../lib/testing'
import { RedwoodSide } from '../side'

const FIXTURE_PATH = path.join(
  __dirname,
  '..',
  '..',
  '..',
  '..',
  '..',
  '__fixtures__'
)
const FIXTURE_PROJECTS = [
  'empty-project',
  'example-todo-main',
  'example-todo-main-with-errors',
  'test-project',
]

describe.each(FIXTURE_PROJECTS)(
  "From within the '%s' fixture",
  (PROJECT_NAME) => {
    // Setup RWJS_CWD to point to the fixture project
    const PROJECT_PATH = path.join(FIXTURE_PATH, PROJECT_NAME)
    const RWJS_CWD = process.env.RWJS_CWD
    beforeAll(() => {
      process.env.RWJS_CWD = PROJECT_PATH
    })
    afterAll(() => {
      process.env.RWJS_CWD = RWJS_CWD
    })

    it('returns the correct sides', () => {
      const sides = RedwoodSide.parseSides()
      expect(sides.length).toMatchSnapshot('sides count')
      for (let i = 0; i < sides.length; i++) {
        const side = sides[i]
        side.filepath = stripAndFormatPathForTesting(
          side.filepath,
          PROJECT_PATH
        )
        expect(side).toMatchSnapshot(`${side.filepath} content`)
        expect(side.getErrors()).toMatchSnapshot(`${side.filepath} errors`)
        expect(side.getWarnings()).toMatchSnapshot(`${side.filepath} warnings`)
      }
    })
  }
)
