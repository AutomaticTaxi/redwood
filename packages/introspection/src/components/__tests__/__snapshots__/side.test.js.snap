// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`From within the 'empty-project' fixture returns the correct sides: /api content 1`] = `
RedwoodSide {
  "filepath": "/api",
  "name": "api",
  "sideType": "api",
  "type": "side",
}
`;

exports[`From within the 'empty-project' fixture returns the correct sides: /api errors 1`] = `[]`;

exports[`From within the 'empty-project' fixture returns the correct sides: /api warnings 1`] = `[]`;

exports[`From within the 'empty-project' fixture returns the correct sides: /web content 1`] = `
RedwoodSide {
  "filepath": "/web",
  "name": "web",
  "sideType": "web",
  "type": "side",
}
`;

exports[`From within the 'empty-project' fixture returns the correct sides: /web errors 1`] = `[]`;

exports[`From within the 'empty-project' fixture returns the correct sides: /web warnings 1`] = `[]`;

exports[`From within the 'empty-project' fixture returns the correct sides: sides count 1`] = `2`;

exports[`From within the 'example-todo-main' fixture returns the correct sides: /api content 1`] = `
RedwoodSide {
  "filepath": "/api",
  "name": "api",
  "sideType": "api",
  "type": "side",
}
`;

exports[`From within the 'example-todo-main' fixture returns the correct sides: /api errors 1`] = `[]`;

exports[`From within the 'example-todo-main' fixture returns the correct sides: /api warnings 1`] = `[]`;

exports[`From within the 'example-todo-main' fixture returns the correct sides: /web content 1`] = `
RedwoodSide {
  "filepath": "/web",
  "name": "web",
  "sideType": "web",
  "type": "side",
}
`;

exports[`From within the 'example-todo-main' fixture returns the correct sides: /web errors 1`] = `[]`;

exports[`From within the 'example-todo-main' fixture returns the correct sides: /web warnings 1`] = `[]`;

exports[`From within the 'example-todo-main' fixture returns the correct sides: sides count 1`] = `2`;

exports[`From within the 'example-todo-main-with-errors' fixture returns the correct sides: /api content 1`] = `
RedwoodSide {
  "filepath": "/api",
  "name": "api",
  "sideType": "api",
  "type": "side",
}
`;

exports[`From within the 'example-todo-main-with-errors' fixture returns the correct sides: /api errors 1`] = `[]`;

exports[`From within the 'example-todo-main-with-errors' fixture returns the correct sides: /api warnings 1`] = `[]`;

exports[`From within the 'example-todo-main-with-errors' fixture returns the correct sides: /web content 1`] = `
RedwoodSide {
  "filepath": "/web",
  "name": "web",
  "sideType": "web",
  "type": "side",
}
`;

exports[`From within the 'example-todo-main-with-errors' fixture returns the correct sides: /web errors 1`] = `[]`;

exports[`From within the 'example-todo-main-with-errors' fixture returns the correct sides: /web warnings 1`] = `[]`;

exports[`From within the 'example-todo-main-with-errors' fixture returns the correct sides: sides count 1`] = `2`;

exports[`From within the 'test-project' fixture returns the correct sides: /api content 1`] = `
RedwoodSide {
  "filepath": "/api",
  "name": "api",
  "sideType": "api",
  "type": "side",
}
`;

exports[`From within the 'test-project' fixture returns the correct sides: /api errors 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct sides: /api warnings 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct sides: /web content 1`] = `
RedwoodSide {
  "filepath": "/web",
  "name": "web",
  "sideType": "web",
  "type": "side",
}
`;

exports[`From within the 'test-project' fixture returns the correct sides: /web errors 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct sides: /web warnings 1`] = `[]`;

exports[`From within the 'test-project' fixture returns the correct sides: sides count 1`] = `2`;
